soroban-sdk-tools-macro/src/error.rs

```diff
@@ -1,30 +1,678 @@
+            },
+            _ => Err(darling::Error::unexpected_lit_type(value)),
+        }
+    }
+
+    fn from_list(items: &[NestedMeta]) -> darling::Result<Self> {
+        let mut mode = ScerrMode::Basic;
+        for item in items {
+            match item {
+                NestedMeta::Meta(Meta::Path(p)) if p.is_ident("root") => {
+                    mode = ScerrMode::Root;
+                }
+                NestedMeta::Meta(Meta::NameValue(nv)) if nv.path.is_ident("mode") => {
+                    mode = match &nv.value {
+                        Expr::Lit(ExprLit { lit: Lit::Str(s), .. }) => match s.value().as_str() {
+                            "root" => ScerrMode::Root,
+                            "basic" => ScerrMode::Basic,
+                            other => return Err(darling::Error::unknown_value(other)),
+                        },
+                        _ => return Err(darling::Error::custom("mode must be a string literal").with_span(&nv.value)),
+                    }
+                }
+                other => return Err(darling::Error::custom("Unsupported #[scerr(...)] argument").with_span(other)),
+            }
+        }
+        Ok(Self(mode))
+    }
+}
+
+fn parse_scerr_attr(attr: TokenStream) -> syn::Result<ScerrMode> {
+    let ts2 = proc_macro2::TokenStream::from(attr);
+    if ts2.is_empty() {
+        return Ok(ScerrMode::Basic);
+    }
+    let meta_items = darling::ast::NestedMeta::parse_meta_list(ts2)
+        .map_err(|e| Error::new(proc_macro2::Span::call_site(), e.to_string()))?;
+    let mode_darling = ScerrModeDarling::from_list(&meta_items)
+        .map_err(|e| Error::new(e.span(), e.to_string()))?;
+    Ok(mode_darling.0)
+}
+
+fn parse_description_attr(attrs: &[Attribute]) -> Option<String> {
+    for attr in attrs {
+        if attr.path().is_ident("description") {
+            if let Ok(meta_nv) = attr.meta.require_name_value() {
+                if let Expr::Lit(ExprLit { lit: Lit::Str(s), .. }) = &meta_nv.value {
+                    return Some(s.value());
+                }
+            } else if let Ok(tokens) = attr.parse_args::<syn::LitStr>() {
+                return Some(tokens.value());
+            }
+        }
+    }
+    None
+}
+
+// -----------------------------------------------------------------------------
+// Variant Processing & Hashing
+// -----------------------------------------------------------------------------
+
+/// DJB2 hash function generating 24-bit codes. Ensures result is never 0.
+fn hash_variant(s: &str) -> u32 {
+    let mut hash = HASH_SEED;
+    for c in s.bytes() {
+        hash = ((hash << 5).wrapping_add(hash)).wrapping_add(c as u32);
+    }
+    let result = hash & INNER_MASK;
+    if result == 0 { 1 } else { result }
+}
+
+#[derive(Debug)]
+struct VariantInfo<'a> {
+    ident: &'a Ident,
+    code: u32,
+    variant: &'a Variant,
+    description: String,
+    is_transparent: bool,
+    is_fcc: bool,
+    is_from_abort: bool,
+    has_from: bool,
```

Comment:
When there are multiple booleans like this it hints that perhaps there should be an enum to describe seens they them to be mutually exclusive.